core/src/main/java/org/springframework/security/oauth2/server/authorization/OAuth2Authorization.java

```diff
@@ -16,16 +16,182 @@
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		OAuth2Authorization that = (OAuth2Authorization) o;
+		return Objects.equals(registeredClientId, that.registeredClientId) &&
+				Objects.equals(principalName, that.principalName) &&
+				Objects.equals(accessToken, that.accessToken) &&
+				Objects.equals(attributes, that.attributes);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(registeredClientId, principalName, accessToken, attributes);
+	}
+
+	@Override
+	public String toString() {
+		return "OAuth2Authorization{" +
+				"registeredClientId='" + registeredClientId + '\'' +
+				", principalName='" + principalName + '\'' +
+				", accessToken=" + accessToken +
+				", attributes=" + attributes +
+				'}';
+	}
+
+	/**
+	 * Returns an empty {@link Builder}.
+	 *
+	 * @return the {@link Builder}
+	 */
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	/**
+	 * Returns a new {@link Builder}, initialized with the provided {@link OAuth2Authorization}.
+	 *
+	 * @param authorization the {@link OAuth2Authorization} to copy from
+	 * @return the {@link Builder}
+	 */
+	public static Builder withOAuth2Authorization(OAuth2Authorization authorization) {
+		Assert.notNull(authorization, "authorization cannot be null");
+		return new Builder(authorization);
+	}
+
+	/**
+	 * Builder class for {@link OAuth2Authorization}.
+	 */
+	public static final class Builder {
+		private String registeredClientId;
+		private String principalName;
+		private OAuth2AccessToken accessToken;
+		private Map<String, Object> attributes;
+
+		protected Builder() {
+		}
+
+		protected Builder(OAuth2Authorization authorization) {
+			this.registeredClientId = authorization.registeredClientId;
+			this.principalName = authorization.principalName;
+			this.accessToken = authorization.accessToken;
+			this.attributes = authorization.attributes;
+		}
+
+		/**
+		 * Sets the registered client identifier.
+		 *
+		 * @param registeredClientId the client id
+		 * @return the {@link Builder}
+		 */
+		public Builder registeredClientId(String registeredClientId) {
+			this.registeredClientId = registeredClientId;
+			return this;
+		}
+
+		/**
+		 * Sets the principal name.
+		 *
+		 * @param principalName the principal name
+		 * @return the {@link Builder}
+		 */
+		public Builder principalName(String principalName) {
+			this.principalName = principalName;
+			return this;
+		}
+
+		/**
+		 * Sets the {@link OAuth2AccessToken}.
+		 *
+		 * @param accessToken the {@link OAuth2AccessToken}
+		 * @return the {@link Builder}
+		 */
+		public Builder accessToken(OAuth2AccessToken accessToken) {
+			this.accessToken = accessToken;
+			return this;
+		}
+
+		/**
+		 * Sets the attributes map.
+		 *
+		 * @param attributes the attributes map
+		 * @return the {@link Builder}
+		 */
+		public Builder attributes(Map<String, Object> attributes) {
```

Comment:
Please remove this method and add:
- `public Builder attribute(String name, String value)`
- `public Builder attributes(Consumer<Map<String, Object>> attributesConsumer)`